Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -170,22 +175,39 @@ ComponentInfo::relativeRotation(const size_t componentIndex) const {
  */
 void ComponentInfo::setPosition(const size_t componentIndex,
                                 const Eigen::Vector3d &newPosition) {
+  // This method is performance critical for some client code. Optimizations are
+  // explained below.
   if (isDetector(componentIndex))
     return m_detectorInfo->setPosition(componentIndex, newPosition);
 
-  scanningCheck(componentIndex);
   const Eigen::Vector3d offset = newPosition - position(componentIndex);
-  const auto indices = this->componentsInSubtree(
-      componentIndex); // Includes requested component index
-  for (auto &index : indices) {
-    if (isDetector(index)) {
-      checkDetectorInfo();
-      m_detectorInfo->setPosition(index,
-                                  m_detectorInfo->position(index) + offset);
-    } else {
-      size_t offsetIndex = compOffsetIndex(index);
-      m_positions.access()[offsetIndex] += offset;
-    }
+
+  // Optimization: Not using detectorsInSubtree and componentsInSubtree to avoid
+  // memory allocations.
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+
+  // Optimization: Split loop over detectors and other components.
+  const auto detRange = (*m_detectorRanges)[rangesIndex];
+  auto it = m_assemblySortedDetectorIndices->begin() + detRange.first;
```

Comment:
I can see how this is faster. My main issue here is what it's done to readability. For both `setPosition` and `setRotation` can you add more comments to each block section. 

We also discussed extracting methods that could return the begin/end iterators, that might improve things slightly.